README.md
# Notification Service Event API Client [![Build Status](https://travis-ci.com/Jimdo/ns-api-event-client-php.svg?token=xaHjcgAFSuULvgxb6q6z&branch=master)](https://travis-ci.com/Jimdo/ns-api-event-client-php)

The client's task is to put events in our *Notification Service SQS queue*. It accomplishes the following things:

  - assemble the event data and create the message body for the SQS request
  - use the SQS client to put the event data into the queue

## Usage

To obtain a client use the factory method:

```
use Jimdo\Notification\Event\Client;

$queueUrl = 'https://sqs.eu-west-1.amazonaws.com/1234567890/ns-api-events';
$awsAccessKey = 'MY_ACCESS_KEY';
$awsSecretKey = 'mys3cr3tk3y';
$awsRegion = 'eu-west-1';

$client = Client::create(
     $queueUrl,
     $awsAccessKey,
     $awsSecretKey,
     $awsRegion
);
```

### To send an event to the SQS queue you have two options

Via the `send()` method:

```
$client->send(
    'website',                // Target of the notification (website or group)
    123456789,                // Target identifier (website id or group descriptor)
    'unique.event.identifier' // A unique event identifier
    'payment.received',       // Namespace of the event
    [                         // Arbitrary payload in key value style
        'color' => 'red',
        'animal' => 'rabbit',
    ]
);
```

Via a *magic* method call:

```
# The method name represents the target of the notification (website or group)

$client->website(
    123456789,                // Website ID to send an event to
    'unique.event.identifier' // A unique event identifier
    'payment.received',       // Namespace of the event
    [                         // Arbitrary payload in key value style
        'color' => 'red',
        'animal' => 'rabbit',
    ]
);
```

**Note:** Although this client is capable of sending events to arbitrary targets the
*Notification Service* currently only supports the target `website`.

## Development

```
# Clone the repo
$ git clone git@github.com:Jimdo/ns-api-event-client-php.git

# Install composer and project dependencies
$ make bootstrap
```

## General Information

This repository provides a `Makefile` to help you speeding up your development process.

```
$ make help
bootstrap    Install composer
tests        Execute test suite and create code coverage report
update       Update composer packages
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.1] - 2017-04-10
### Fixed
- Don't enforce payload to be an json object. 
Client can enforce empty array to be an object by explicitly type casting, for example: ```(object)[]``` from @alexzelenuyk 

## [0.3.0] - 2017-03-16
### Changed
- write events directly into the SQS queue from @partyschaum

## [0.2.1] - 2017-02-10
### Changed
- remove key value pairs with empty value in nested keys in payload from @sbstjn

## [0.2.0] - 2017-02-08
### Changed
- remove key value pairs with empty strings from payload from @partyschaum

## 0.1.0 - 2016-12-15
### Added
- run tests in docker container: `make tests` from @alexzelenuyk
- run `composer install` in docker container: `make bootstrap` from @alexzelenuyk
- run `composer update` in docker container: `make update` from @alexzelenuyk

[Unreleased]: https://github.com/Jimdo/ns-api-event-client-php/compare/0.3.1...HEAD
[0.3.1]: https://github.com/Jimdo/ns-api-event-client-php/compare/0.3.0...0.3.1
[0.3.0]: https://github.com/Jimdo/ns-api-event-client-php/compare/0.2.1...0.3.0
[0.2.1]: https://github.com/Jimdo/ns-api-event-client-php/compare/0.2.0...0.2.1
[0.2.0]: https://github.com/Jimdo/ns-api-event-client-php/compare/0.1.0...0.2.0


tests/ClientTest.php
<?php

namespace Jimdo\Notification\Event;

use PHPUnit\Framework\TestCase;

use GuzzleHttp\Client as HttpClient;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\Middleware;
use GuzzleHttp\Psr7\Response;

use Aws\Credentials\Credentials;
use Aws\Sqs\SqsClient;

class ClientTest extends TestCase
{
    private $requests;

    private $cannedResponseForAllSqsRequests = '<?xml version="1.0"?><SendMessageResponse xmlns="http://queue.amazonaws.com/doc/2012-11-05/"><SendMessageResult><MessageId>95809e0e-2cf7-44cc-bf44-1d9ffbe7ec3c</MessageId><MD5OfMessageBody>961770dca2cb59402568f975513228fe</MD5OfMessageBody></SendMessageResult><ResponseMetadata><RequestId>19293607-af0f-596d-806d-6744cd17de91</RequestId></ResponseMetadata></SendMessageResponse>';

    protected function setUp()
    {
        $this->requests = [];
    }

    /**
     * @test
     */
    public function itShouldRemoveEmptyStringValuesFromPayload()
    {
        $created = (int)(microtime(true) * 1000);

        $client = $this->client(['timeInMilliseconds']);
        $client->method('timeInMilliseconds')->willReturn($created);

        $client->send(
            'website',
            1234567,
            'unique.event.identifier',
            'payment.received',
            [
                'color' => 'red',
                'keyWithEmptyValue' => '',
                'number' => 0
            ]
        );

        $requestBody = $this->getMessageBody();

        $this->assertNotContains('keyWithEmptyValue', $requestBody);
    }

    /**
     * @test
     */
    public function itShouldRemoveEmptyStringValuesFromPayloadInChildArrays()
    {
        $created = (int)(microtime(true) * 1000);

        $client = $this->client(['timeInMilliseconds']);
        $client->method('timeInMilliseconds')->willReturn($created);

        $client->send(
            'website',
            1234567,
            'unique.event.identifier',
            'payment.received',
            [
                'color' => 'red',
                'keyWithEmptyChild' => [
                  'keyWithEmptyValue' => ''
                ],
                'keyWithEmptyChildInChild' => [
                  'keyWithEmptyChild' => [
                    'secondKeyWithEmptyValue' => ''
                  ],
                ],
                'keyWithEmptyArray' => [],
                'number' => 0
            ]
        );

        $requestBody = $this->getMessageBody();

        $this->assertContains('color', $requestBody);
        $this->assertContains('number', $requestBody);

        $this->assertNotContains('keyWithEmptyValue', $requestBody);
        $this->assertNotContains('secondKeyWithEmptyValue', $requestBody);
        $this->assertNotContains('keyWithEmptyArray', $requestBody);
        $this->assertNotContains('keyWithEmptyChild', $requestBody);
        $this->assertNotContains('keyWithEmptyChildInChild', $requestBody);
    }

    /**
     * @test
     */
    public function itShouldGenerateCorrectRequestBody()
    {
        $created = (int)(microtime(true) * 1000);

        $client = $this->client(['timeInMilliseconds']);
        $client->method('timeInMilliseconds')->willReturn($created);

        $client->send(
            'website',
            12345566,
            'unique.event.identifier',
            'payment.received',
            [
                'color' => 'red',
                'animal' => 'rabbit',
            ]
        );

        $requestBody = $this->getMessageBody();

        $expectedRequestBody = '{"id":"unique.event.identifier","name":' .
            '"payment.received","website":12345566,"created":' . $created .
            ',"payload":{"color":"red","animal":"rabbit"}}';

        $this->assertEquals($expectedRequestBody, $requestBody);
    }

    /**
     * @test
     */
    public function itShouldGenerateAlwaysIdOfTypeString()
    {
        $created =  (int)(microtime(true) * 1000);

        $client = $this->client(['timeInMilliSeconds']);
        $client->method('timeInMilliSeconds')->willReturn($created);

        $client->send(
            'website',
            12345566,
            123,
            'payment.received',
            [
                'color' => 'red',
                'animal' => 'rabbit',
            ]
        );

        $requestBody = $this->getMessageBody();

        $expectedRequestBody = '{"id":"123","name":' .
            '"payment.received","website":12345566,"created":' . $created .
            ',"payload":{"color":"red","animal":"rabbit"}}';

        $this->assertEquals($expectedRequestBody, $requestBody);
    }

    /**
     * @test
     */
    public function itShouldGenerateAlwaysJSONObjectFromEmptyPayload()
    {
        $created =  (int)(microtime(true) * 1000);

        $client = $this->client(['timeInMilliSeconds']);
        $client->method('timeInMilliSeconds')->willReturn($created);

        $client->send(
            'website',
            12345566,
            123,
            'payment.received',
            []
        );

        $requestBody = $this->getMessageBody();

        $expectedRequestBody = '{"id":"123","name":' .
            '"payment.received","website":12345566,"created":' . $created .
            ',"payload":{}}';

        $this->assertEquals($expectedRequestBody, $requestBody);
    }

    /**
     * @test
     */
    public function itShouldGenerateCorrectJsonObjectFromPayloadWithMixedContent()
    {
        $created =  (int)(microtime(true) * 1000);

        $client = $this->client(['timeInMilliSeconds']);
        $client->method('timeInMilliSeconds')->willReturn($created);

        $client->send(
            'website',
            12345566,
            123,
            'payment.received',
            [
                'array' => ['item1', 'item2'],
                'object' => (object)['item1', 'item2'],
            ]
        );

        $requestBody = $this->getMessageBody();

        $expectedRequestBody = '{"id":"123","name":' .
            '"payment.received","website":12345566,"created":' . $created .
            ',"payload":{"array":["item1","item2"],"object":{"0":"item1","1":"item2"}}}';

        $this->assertEquals($expectedRequestBody, $requestBody);
    }

    /**
     * @test
     */
    public function itShouldGenerateRequestForCorrectQueueUrlMagically()
    {
        $created = (int)(microtime(true) * 1000);

        $client = $this->client(['timeInMilliseconds']);
        $client->method('timeInMilliseconds')->willReturn($created);

        $client->website(
            12345566,
            'unique.event.identifier',
            'payment.received',
            [
                'color' => 'red',
                'animal' => 'rabbit',
            ]
        );

        $requestBody = $this->getMessageBody();

        $expectedRequestBody = '{"id":"unique.event.identifier","name":' .
            '"payment.received","website":12345566,"created":' . $created .
            ',"payload":{"color":"red","animal":"rabbit"}}';

        $this->assertEquals($expectedRequestBody, $requestBody);
    }

    /**
     * @param array $methodsToStub
     * @return Client|PHPUnit_Framework_MockObject_MockObject
     */
    private function client($methodsToStub)
    {
        return $this->getMockBuilder('Jimdo\Notification\Event\Client')
            ->setConstructorArgs([$this->config(), $this->sqsClient()])
            ->setMethods($methodsToStub)
            ->getMock();
    }

    /**
     * @return Client\Config
     */
    private function config()
    {
        $queueUrl = 'https://sqs.eu-west-1.amazonaws.com/1234567890/ns-api-events';
        $awsKey = 'AWS_ACCESS_KEY';
        $awsSecret = 's0m3aw5s3cr31';
        $awsRegion = 'eu-west-1';

        return new Client\Config(
            $queueUrl,
            $awsKey,
            $awsSecret,
            $awsRegion
        );
    }

    /**
     * @return \GuzzleHttp\Client
     */
    private function handlers()
    {
        $stack = HandlerStack::create(new MockHandler([
            new Response(200, [], $this->cannedResponseForAllSqsRequests)
        ]));

        $history = Middleware::history($this->requests);
        $stack->push($history);

        return $stack;
    }

    /**
     * @return \Aws\Sqs\SqsClient
     */
    private function sqsClient()
    {
        return  new SqsClient([
            'credentials' => new Credentials($this->config()->key(), $this->config()->secret()),
            'region' => $this->config()->region(),
            'version' => Client::SQS_API_VERSION,
            'http_handler' => $this->handlers(),
        ]);
    }

    /**
     * @return string
     */
    private function getMessageBody()
    {
        $requestBody = (string) $this->requests[0]['request']->getBody();
        $queryParams = [];
        parse_str(urldecode($requestBody), $queryParams);
        return $queryParams['MessageBody'];
    }
}


tests/DummySignature.php
<?php

namespace Jimdo\Notification\Event;

use Aws\Signature\SignatureInterface;
use Aws\Credentials\CredentialsInterface;
use Psr\Http\Message\RequestInterface;

class DummySignature implements SignatureInterface
{
    /**
     * Signs the specified request with an AWS signing protocol by using the
     * provided AWS account credentials and adding the required headers to the
     * request.
     *
     * @param RequestInterface     $request     Request to sign
     * @param CredentialsInterface $credentials Signing credentials
     *
     * @return RequestInterface Returns the modified request.
     */
    public function signRequest(RequestInterface $request, CredentialsInterface $credentials)
    {
        return $request;
    }

    /**
     * Create a pre-signed request.
     *
     * @param RequestInterface     $request     Request to sign
     * @param CredentialsInterface $credentials Credentials used to sign
     * @param int|string|\DateTime $expires The time at which the URL should
     *     expire. This can be a Unix timestamp, a PHP DateTime object, or a
     *     string that can be evaluated by strtotime.
     *
     * @return RequestInterface
     */
    public function presign(RequestInterface $request, CredentialsInterface $credentials, $expires)
    {
        return $request;
    }
}


src/Client/Config.php
<?php

namespace Jimdo\Notification\Event\Client;

class Config
{
    /** @var string */
    private $queueUrl;

    /** @var array */
    private $awsKey;

    /** @var string */
    private $awsSecret;

    /** @var string */
    private $awsRegion;

    /**
     * @param string $queueUrl
     * @param string $awsKey
     * @param string $awsSecret
     * @param string $awsRegion
     */
    public function __construct($queueUrl, $awsKey, $awsSecret, $awsRegion)
    {
        $this->queueUrl = $queueUrl;
        $this->awsKey = $awsKey;
        $this->awsSecret = $awsSecret;
        $this->awsRegion = $awsRegion;
    }

    /**
     * @return string
     */
    public function queueUrl()
    {
        return $this->queueUrl;
    }

    /**
     * @return string AWS key
     */
    public function key()
    {
        return $this->awsKey;
    }

    /**
     * @return string AWS secret
     */
    public function secret()
    {
        return $this->awsSecret;
    }

    /**
     * @return string AWS region
     */
    public function region()
    {
        return $this->awsRegion;
    }
}


src/Client.php
<?php

namespace Jimdo\Notification\Event;

use Aws\Credentials\Credentials;
use Aws\Sqs\SqsClient;

use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Client as HttpClient;

class Client
{
    const SQS_API_VERSION = '2012-11-05';

    /** @var Client\Config */
    private $config;

    /** @var \Aws\Sqs\SqsClient */
    private $sqsClient;

    /**
     * @param Client\Config $config
     * @param \Aws\Sqs\SqsClient $sqsClient
     */
    public function __construct(Client\Config $config, \Aws\Sqs\SqsClient $sqsClient)
    {
        $this->config = $config;
        $this->sqsClient = $sqsClient;
    }

    /**
     * @param string $queueUrl
     * @param string $awsKey
     * @param string $awsSecret
     * @param string $awsRegion
     * @param \Aws\Sqs\SqsClient $sqsClient
     * @return Client
     */
    public static function create($queueUrl, $awsKey, $awsSecret, $awsRegion, $sqsClient = null)
    {
        if ($sqsClient === null) {
            $sqsClient = new SqsClient([
                'credentials' => new Credentials($awsKey, $awsSecret),
                'region' => $awsRegion,
                'version' => self::SQS_API_VERSION,
            ]);
        }

        return new self(
            new Client\Config($queueUrl, $awsKey, $awsSecret, $awsRegion),
            $sqsClient
        );
    }

    /**
     * @param string $target Target of the notification
     * @param array $arguments As in send() but without $target
     * @return \Psr\Http\Message\ResponseInterface
     */
    public function __call($target, $arguments)
    {
        list($targetId, $id, $name, $payload) = $arguments;
        return $this->send($target, $targetId, $id, $name, $payload);
    }

    /**
     * @param string $target Target of the notification
     * @param mixed $targetId Target identifier (website id or group descriptor)
     * @param string $id Unique event identifier
     * @param string $name Event name
     * @param array $payload Arbitrary JSON payload
     * @return \Aws\Result
     */
    public function send($target, $targetId, $id, $name, array $payload)
    {
        return $this->sqsClient->sendMessage([
            'QueueUrl' => $this->config->queueUrl(),
            'MessageBody' => $this->messageBody(
                $target,
                $targetId,
                $id,
                $name,
                $this->removeEmptyKeys($payload)
            ),
        ]);
    }

    /**
     * @param string $target Path fragment to lambda function
     * @param mixed $targetId Target identifier (website id or group descriptor)
     * @param string $id Unique event identifier
     * @param string $name Event name
     * @param array $payload Arbitrary JSON payload
     * @return \GuzzleHttp\Psr7\Request
     */
    private function messageBody($target, $targetId, $id, $name, array $payload)
    {
        return json_encode([
            'id' => (string) $id,
            'name' => $name,
            $target => $targetId,
            'created' => $this->timeInMilliseconds(),
            'payload' => empty($payload) ? (object)[] : $payload,
        ]);
    }

    /**
     * @param arary $array
     * @return array Filtered array
     */
    private function removeEmptyKeys($array) {
        $callback = function($item) use (&$callback) {
            if (is_array($item)) {
                return array_filter($item, $callback);
            }

            return is_numeric($item) || !empty($item);
        };

        return array_filter($array, $callback);
    }

    /**
     * @return int
     */
    protected function timeInMilliseconds()
    {
        return (int)(microtime(true) * 1000);
    }
}
